hawkular-bus-common/src/main/java/org/hawkular/bus/common/BasicMessage.java

```diff
@@ -16,243 +16,27 @@
-            final int released = parser.releaseBuffered(remainderStream);
-            remainder = (released > 0) ? remainderStream.toByteArray() : new byte[0];
-        } catch (Exception e) {
-            throw new IllegalArgumentException("Stream cannot be converted to JSON object of type [" + clazz + "]", e);
-        }
-        return new BasicMessageWithExtraData<T>(obj, new BinaryData(remainder, in));
-    }
-
-    private static Method findBuildObjectMapperForDeserializationMethod(Class<? extends BasicMessage> clazz) {
-        try {
-            Method m = clazz.getDeclaredMethod("buildObjectMapperForDeserialization");
-            return m;
-        } catch (Exception e) {
-            // the given subclass doesn't have a method to build a mapper, maybe its superclass does.
-            // eventually we'll get to the BasicMessage class and we know it does have one.
-            return findBuildObjectMapperForDeserializationMethod((Class<? extends BasicMessage>) clazz.getSuperclass());
-        }
-    }
-
-    /**
-     * This is static, so really there is no true overriding it in subclasses.
-     * However, fromJSON will do the proper reflection in order to invoke
-     * the proper method for the class that is being deserialized. So subclasses
-     * that want to provide their own ObjectMapper will define their own
-     * method that matches the signature of this method and it will be used.
-     *
-     * @return object mapper to be used for deserializing JSON.
-     */
-    protected static ObjectMapper buildObjectMapperForDeserialization() {
-        final ObjectMapper mapper = new ObjectMapper();
-        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
-        return mapper;
-    }
-
-    /**
-     * Converts this message to its JSON string representation.
-     *
-     * @return JSON encoded data that represents this message.
-     */
-    public String toJSON() {
-        final ObjectMapper mapper = buildObjectMapperForSerialization();
-        try {
-            return mapper.writeValueAsString(this);
-        } catch (JsonProcessingException e) {
-            throw new IllegalStateException("Object cannot be parsed as JSON.", e);
-        }
-    }
+public interface BasicMessage {
 
-    /**
-     * @return object mapper to be used to serialize a message to JSON
-     */
-    protected ObjectMapper buildObjectMapperForSerialization() {
-        final ObjectMapper mapper = new ObjectMapper();
-        mapper.setVisibilityChecker(mapper.getSerializationConfig().getDefaultVisibilityChecker()
-                .withFieldVisibility(JsonAutoDetect.Visibility.ANY)
-                .withGetterVisibility(JsonAutoDetect.Visibility.NONE)
-                .withSetterVisibility(JsonAutoDetect.Visibility.NONE)
-                .withCreatorVisibility(JsonAutoDetect.Visibility.NONE));
-        return mapper;
-    }
+    String HEADER_BASIC_MESSAGE_CLASS = "org.hawkular.bus.common.BasicMessage.className";
```

Comment:
In my opinion, this constant should not be here. It doesn't have anything to do with the messages per se. This has to do with processing messages over the bus. A more appropriate place to define this constant is in some place like MessageProcessor (since that is where the headers are added).